src/test/java/org/folio/circulation/support/fetching/PageableFetcherTest.java

```diff
@@ -0,0 +1,123 @@
+    final var pageProcessor = spy(dummyProcessor());
+
+    final var voidResult = processPages(repository, pageSize, pageProcessor);
+
+    assertThat(voidResult.succeeded(), is(true));
+    verify(pageProcessor, times(10)).processPage(any());
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldAbortFetchingWhenPageProcessorRaisedError() {
+    final var failureMessage = "Fetch failure";
+    final var pageSize = limit(10);
+    final var repository = repository(50);
+    final PageProcessor<Integer> pageProcessor = mock(PageProcessor.class);
+
+    when(pageProcessor.processPage(any()))
+      .thenReturn(completedFuture(failed(new ServerErrorFailure(failureMessage))));
+
+    final var voidResult = processPages(repository, pageSize, pageProcessor);
+
+    assertThat(voidResult.failed(), is(true));
+    assertThat(voidResult.cause().toString(), containsString(failureMessage));
+    verify(pageProcessor, times(1)).processPage(any());
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldAbortFetchingWhenFailedToFetch() {
+    final var failureMessage = "Fetch failure";
+    final var pageSize = limit(10);
+    final GetManyRecordsRepository<Integer> repository = mock(GetManyRecordsRepository.class);
+    final PageProcessor<Integer> pageProcessor = mock(PageProcessor.class);
+
+    when(repository.getMany(any(), any(), any()))
+      .thenReturn(completedFuture(failed(new ServerErrorFailure(failureMessage))));
+
+    final var voidResult = processPages(repository, pageSize, pageProcessor);
+
+    assertThat(voidResult.failed(), is(true));
+    assertThat(voidResult.cause().toString(), containsString(failureMessage));
+    verify(pageProcessor, noInteractions()).processPage(any());
+  }
+
+  @Test
+  public void shouldAbortFetchingWhenRecordCountLimitIsReached() {
+    final var pageSize = limit(10);
+    final var recordLimit = 90;
+    final GetManyRecordsRepository<Integer> repository = repository(recordLimit + 10);
+    final PageProcessor<Integer> pageProcessor = dummyProcessor();
+
+    final var voidResult = new PageableFetcher<>(repository, pageSize, recordLimit)
+      .processPages(noQuery().value(), pageProcessor)
+      .getNow(Result.failed(new ServerErrorFailure("Time out")));
+
+    assertThat(voidResult.failed(), is(true));
+    assertThat(voidResult.cause().toString(), containsString(
+      "Maximum allowed item count is set to 90 and it has been reached"));
+  }
+
+  private <T> Result<Void> processPages(GetManyRecordsRepository<T> repository,
+    PageLimit pageLimit, PageProcessor<T> processor) {
+
+    return new PageableFetcher<>(repository, pageLimit, 1000)
+      .processPages(noQuery().value(), processor)
+      .getNow(Result.failed(new ServerErrorFailure("Time out")));
+  }
+
+  private GetManyRecordsRepository<Integer> repository(int totalRecords) {
+    return (cqlQuery, pageLimit, offset) -> {
+      final var start = offset.getOffset() + 1;
```

Comment:
Why is start one more than offset?

Reply:
Removed this +1. Now it starts from offset value.